File: src/api/bookmark.ts
import { call } from './tauri'

export type Bookmark = {
  id: string
  bookId: string
  chapterId: string
  pageIndex: number
  title: string
  note?: string | null
  createdAt: string
  updatedAt: string
}

export type CreateBookmarkRequest = {
  bookId: string
  chapterId: string
  pageIndex: number
  title?: string | null
  note?: string | null
}

export function listBookmarks(bookId: string): Promise<Bookmark[]> {
  return call('list_bookmarks', { bookId })
}

export function createBookmark(request: CreateBookmarkRequest): Promise<Bookmark> {
  return call('create_bookmark', { request })
}

export function deleteBookmark(bookmarkId: string): Promise<void> {
  return call('delete_bookmark', { bookmarkId })
}

export function isPageBookmarked(bookId: string, chapterId: string, pageIndex: number): Promise<boolean> {
  return call('is_page_bookmarked', { bookId, chapterId, pageIndex })
}

File: src-tauri/src/commands/bookmark.rs
use tauri::State;

use crate::{
    errors::AppResult,
    models::bookmark::{Bookmark, CreateBookmarkRequest},
    AppState,
};

#[tauri::command]
pub fn list_bookmarks(state: State<'_, AppState>, book_id: String) -> AppResult<Vec<Bookmark>> {
    state.database.list_bookmarks(&book_id)
}

#[tauri::command]
pub fn create_bookmark(
    state: State<'_, AppState>,
    request: CreateBookmarkRequest,
) -> AppResult<Bookmark> {
    state.database.create_bookmark(&request)
}

#[tauri::command]
pub fn delete_bookmark(state: State<'_, AppState>, bookmark_id: String) -> AppResult<()> {
    state.database.delete_bookmark(&bookmark_id)
}

#[tauri::command]
pub fn is_page_bookmarked(
    state: State<'_, AppState>,
    book_id: String,
    chapter_id: String,
    page_index: usize,
) -> AppResult<bool> {
    state
        .database
        .is_page_bookmarked(&book_id, &chapter_id, page_index)
}
